CAT/strategy/NCAT_strategy.py

```diff
@@ -0,0 +1,665 @@
+        if mask is not None:
+            # Same mask applied to all h heads.
+            mask = mask.unsqueeze(1)
+        nbatches = query.size(0)
+        
+        # 1) Do all the linear projections in batch from d_model => h x d_k 
+        query, key, value = \
+            [l(x).view(nbatches, -1, self.h, self.d_k).transpose(1, 2)
+             for l, x in zip(self.linears, (query, key, value))]
+
+        # 2) Apply attention on all the projected vectors in batch. 
+        _, x, self.attn = attention(query, key, value, mask=mask, 
+                                 dropout=self.dropout)
+        
+        # 3) "Concat" using a view and apply a final linear. 
+        x = x.transpose(1, 2).contiguous() \
+             .view(nbatches, -1, self.h * self.d_k)
+        return self.linears[-1](x)
+
+class MultiHeadedAttention_con(nn.Module):
+    def __init__(self, h, d_model, dropout=0.1):
+        "Take in model size and number of heads."
+        super(MultiHeadedAttention_con, self).__init__()
+        assert d_model % h == 0
+        # We assume d_v always equals d_k
+        self.d_k = d_model // h
+        self.h = h
+        self.linears = clones(nn.Linear(d_model, d_model), 5)
+        self.dropout = nn.Dropout(p=dropout)
+        
+    def forward(self, query, key, value1, value2, mask=None):
+        if mask is not None:
+            # Same mask applied to all h heads.
+            mask = mask.unsqueeze(1)
+        nbatches = query.size(0)
+        
+        # 1) Do all the linear projections in batch from d_model => h x d_k 
+        query, key = \
+            [l(x).view(nbatches, -1, self.h, self.d_k).transpose(1, 2)
+             for l, x in zip(self.linears, (query, key))]
+
+        value1 = value1.view(nbatches, -1, self.h, self.d_k).transpose(1, 2)
+        value2 = value2.view(nbatches, -1, self.h, self.d_k).transpose(1, 2)
+        # 2) Apply attention on all the projected vectors in batch. 
+        
+        # print(query.shape, key.shape, value1.shape)
+        scores, x1, attn_s = attention(query, key, value1, mask=mask, 
+                                 dropout=self.dropout)
+
+        x2 = torch.matmul(attn_s.transpose(-1, -2), value2)
+        
+        # 3) "Concat" using a view and apply a final linear. 
+        x1 = x1.transpose(1, 2).contiguous() \
+             .view(nbatches, -1, self.h * self.d_k)
+        
+        x2 = x2.transpose(1, 2).contiguous() \
+             .view(nbatches, -1, self.h * self.d_k)
+        return self.linears[-1](x1), self.linears[-1](x2), 
+
+class PositionwiseFeedForward(nn.Module):
+    "Implements FFN equation."
+    def __init__(self, d_model, d_ff, dropout=0.1):
+        super(PositionwiseFeedForward, self).__init__()
+        self.w_1 = nn.Linear(d_model, d_ff)
+        self.w_2 = nn.Linear(d_ff, d_model)
+        self.dropout = nn.Dropout(dropout)
+
+    def forward(self, x):
+        return self.w_2(self.dropout(F.relu(self.w_1(x))))
+    
+class IRT(nn.Module):
```

Comment:
Remove IRT from this file. Make sure to implement all CDMs in /model and only implement the question selection strategy in /strategy